### PR TITLE
Implement CSS Grid

### LINKED ISSUE #204: Support CSS Grid
As per https://github.com/DioxusLabs/taffy/issues/28#issuecomment-1179532564, we would like to support CSS Grid in Taffy. This is a modified version of research I did on supporting this in Stretch (https://github.com/vislyhq/stretch/issues/63#issuecomment-616121239). Which I thought I would copy and update here as a starting point.

#### Status (22nd Dec 2022)

The implementation is largely feature complete, and a [pre-release version](https://github.com/DioxusLabs/taffy/blob/main/RELEASES.md#030-alpha1) of Taffy containing the CSS Grid implementation has been released. There is now a lot of testing and validating to be done before it is ready for a stable release. 

### Blocking questions / issues
- [x] ~How should Grid interact with intrinsically sized leaf nodes. The spec seems to define child nodes as having both a `min-size` (e.g. text node that wraps at all opportunities) and `max-size` (e.g. a text node that doesn't wrap at all). Possibly these would need to be added to nodes as additional  "measure functions"? Or possibly they could both be assumed to be equal to the one measure functions at first? Possibly min can be obtained by passing the MeasureFunc `width: 0, height: 0`, and max can be obtained by passing the MeasureFunc `width: undefined, height: undefined`? But perhaps it would be better to redefine the MeasureFunc to "sizing constraints" (https://www.w3.org/TR/css-sizing-3/#constraints) in addition a definite size constraint if one is available? https://www.w3.org/TR/css-sizing-3/#intrinsic-contribution~
Plan is to update `MeasureFunc` to take the sizing constraint as parameter (which it can then use or ignore as it sees fit).
- [x] ~The interaction between nested Grid and Flexbox nodes. The comment at https://github.com/DioxusLabs/taffy/blob/main/src/flexbox.rs#L463-L466 suggests that the Flex implementation could be made more spec compliant by passing min-content/max-content constraints into to the top (presumably of the `compute_internal` function. I feel like this may be a prerequisite to getting Grid and Flex to play nicely with each other in indefinitely sized contexts. Tracking issue here: https://github.com/DioxusLabs/taffy/issues/122~
Plan is not to worry about integration at first, but it's likely that we'll adjust the Flexbox implementation to take sizing constraints as input as suggested above when we are ready to integrate.

**Update**: Both of these solutions were implemented as part of #246. Grid's integration with other layout modes is now unblocked.

### Useful Resources:

- CSS Grid Level 1 Specification: https://www.w3.org/TR/css-grid-1
- CSS Grid Level 2 Specification: https://www.w3.org/TR/css-grid-2
- CSS Sizing Level 3 Sizing Values https://www.w3.org/TR/css-sizing-3/#sizing-values
- CSS Sizing Level 3  Intrinsic Sizes https://www.w3.org/TR/css-sizing-3/#intrinsic-sizes
- CSS Alignment Level 3 https://www.w3.org/TR/css-align-3/
- MDN Explanation of Subgrid (https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout/Subgrid)
- CSS Grid Garden: https://cssgridgarden.com
- CSS Tricks [Complete guide to CSS Grid](https://css-tricks.com/snippets/css/complete-guide-grid/#aa-properties-for-the-grid-container)
- MDN value definition syntax https://developer.mozilla.org/en-US/docs/Web/CSS/Value_definition_syntax

### MVP Implementation Plan

- [x] Extend the Style and Geometry sections of Stretch to include definitions for CSS Grid styles, while ensuring that the existing Flex implementation continues to work:
    - [x] ~`fr` units~ This has been implemented as the `Flex` variant of the new `TrackSizingFunction` enum rather than being added to `Dimension` as it only applies to grid templates, and it turns out there are also a bunch of other options (like `minmax`) that apply there but not elsewhere.
    - [x] `display: grid`
    - [x] `grid-template-columns`/`grid-template-rows`
    - [x] `grid-auto-rows`/`grid-auto-columns`/`grid-auto-flow`
    - [x] `grid-column` / `grid-row`/ `grid-row-start`/`grid-row-end`/`grid-column-start`/`grid-column-end`
    - [x] `row-gap`/`column-gap` (note: `gap` shorthand implemented instead)
    - [x]  `justify-items` / `justify-self` (`justify-content` / `align-content` / `align-items` / `align-self` already implemented for flexbox). Note: \*-items and \*-self properties align grid items (children) within their cell. \*-content properties align cells (rows/columns) within the grid container. I *think* that all are set to stretch to fit by default.
    
    We can probably ignore the shorthand properties for now, as their functionality is covered by the longhand equivalents above.

- [x] Resolve the explicit grid:
  - [x] Count the number of tracks in the "explicit grid". This is the rows and columns as defined by `grid-template-columns`/`grid-template-rows` (Sections 7.1-7.4 of the spec).
  - [x] Extend gentest infrastructure to support `repeat()` syntax
  - [x] Implement resolving fixed repetition explicit tracks `repeat(integer)` #342
  - [x] Implement resolving auto-generated explicit tracks `repeat(auto-fill)`
  - [x] Implement collapsing tracks for `repeat(auto-fit)`
  
  (Not strictly necessary features like named grid lines/areas defined by `grid-template-areas` will be left until later.)

- [x] Implement grid placement. Matching Grid Items (child nodes of the grid) to Grid Areas. This may include generating extra rows and columns that form the "implicit grid" if any Grid Items are placed outside of the bounds of the "explicit grid". (note: the "implicit grid" is the whole grid, not just the extra bit) (Sections 7.5-8.5 of the spec)

- [x] Resolve the track sizing functions for all grid tracks (explicit and implicit).

- [x] Implement the Grid Sizing Algorithm to determine the size of each Grid Area. This mostly consists of the Track Sizing Algorithm, iterated a few times. A Grid Track is a row or column of the grid. Gutters sizes should also be resolved here, although this is relatively simple. (Section 11 of the spec).
  - [x] Resolve size of fixed size tracks (11.4. "Initialize Track Sizes" and 11.6. "Maximise Tracks")
  - [x] Resolve size of flexible tracks (with no children affecting size) (11.7 "Expand Flexible Tracks")
  - [x] Resolve size of intrinsically sized tracks (with no children affecting size) (11.8. "Stretch auto Tracks")
  - [x] Allow children of span 1 to affect track size (11.5. "Resolve Intrinsic Track Sizes" Step 2)
  - [x] Allow children with span > 1 to affect track size (11.5. "Resolve Intrinsic Track Sizes" Step 3 and 4)
  - [x] Support for `fit-content()` track sizing functions
    - [x] Fixed points fit-content arguments
    - [x] Percentage fit-content arguments

- [x] Resolve Grid Item (node) sizes and positions, by summing the size of the track and gutters in their grid area (in each axis), and taking into accounts spacing and alignment properties (Section 10 of the spec).
   - [x] Align grid track using `align_content` and `justify_content`
   - [x] Align items within their grid area using `align_items`/`align_self`/`justify_items`/`justify_self`

- [x] Resolve sizes and positions of absolutely positioned elements (Section 9 of the spec)
- [x] Support for baseline alignment #344

### Code quality tasks:
  - [x] Fix all warnings
  - [x] Fix clippy lints
  - [x] Convert track sizing function enums to use `LengthPercentage` rather than `Dimension`
  - [x] Implement existing style helper functions (auto, zero, points) for grid styles
  - [x] Add new style helpers:
     - [x] percent
     - [x] min_content
     - [x] max_content
     - [x] flex
     - [x] minmax
     - [x] repeat
     - [x] track (line?)
     - [x] span
  - [x] Refactor code that deals with axes (make consistent)
  - [x] Evaluate use of `grid` crate
  - [x] Update release notes to reflect the changes in this branch
    - [x] Updates to alignment styles
    - [x] CSS Grid layout mode
  - [x] Add feature flag for CSS grid feature (`experimental_grid`?)
  
### Correctness fixes:
  - [x] Don't distribute maximum's space to growth limits for flex items
  - [x] Apply align-content/justify-content when estimating opposite-axis available space
  - [x] Crash when children are implicitly placed outside the explicit grid
  - [x] More careful handling of grid lines indexes. One should not do arithmetic on grid lines without using the explicit track count as it is non-linear around 0. Grid line zero should be treated as "auto" and/or `GridPlacement::Line` should use `NonZeroI16`. #301
  - [x] #346
  - [x] Distribute space in proportion to flexible sizing functions #349 
  
### To test thoroughly
   - [x] ~Repetition (`repeat(integer)`, `repeat(auto-fill)` and `repeat(auto-fit)`)~ (covered by unit tests. could still do with gentests but lower priority)
   - [x] ~Placement using Span~ (covered by unit tests. could still do with gentests but lower priority)
   - [x] ~Flex-column children~ We don't match chrome on this, but our behaviour (actually containing the width of the child node) seems better, and matches the spec as best as I can tell.
   - [x] Containers with padding/borders
   - [x] Absolutely positioned children
   - [x] Hidden children
   - [x] Children with margins
   - [x] Percent-size grid containers sizing into definite available_space #306
   - [x] Children with percentage width/height/padding/border/margin #306
   - [x] Top-to-bottom writing direction children #300
   - [x] Percentage sized tracks #314
   - [x] Children with defined aspect ratio #317
   - [x] Min and max width/height #317
   - [x] Fit-content (including percentage arguments) #335
   - [x] Items spanning both fixed and percentage gaps #335
   - [x] Baseline Alignment #344
   - [x] Distribute space in proportion to flexible sizing functions #349
   - [x] Items spanning multiple tracks where the tracks are flexible or intrinsically sized #349
   - [x] Content sizing of `minmax()` tracks #349
   - [x] Tracks with `fr` values summing to less than 1 #349
   - [x] Complex integration tests (including Interaction with Flexbox):
      - [x] https://github.com/DioxusLabs/taffy/pull/354
      - [x] https://github.com/DioxusLabs/taffy/pull/355
   
### Performance related tasks:

- [x] #340 
- [x] Fix exponential blowup with regard to nesting depth #338
- [x] Add tests checking for relayout counts #338
- [x] Only compute child intrinsic size when needed #338
- [x] Add early exit for grid container measurement #338
- [x] Only do second track sizing iteration if "any of the item's min-content contributions have changed" #341 
- [x] Implement optimised track sizing for items that span exactly 1 track #343 

#### Grid related further tasks

- Support for named grid lines
- Support for `grid-template-areas` and `grid-area`
- Support for CSS Grid Level 2 (subgrids)

#### Non-grid further tasks
- Support for `min-content`/`max-content`/`fit-content` sizes (widths/heights)
- Add sub-directory support to gentests to separate grid and flexbox tests.

### PR DESCRIPTION
**Pull request by nicoburns**
# Objective

Fixes #204

## Feedback wanted

Nothing in particular just yet. This PR is still very incomplete. Just putting it up early to track progress.

### CODE REVIEW

**Comment by TimJentzsch** [comment]
FYI, if your work is still in progress it's good practice to mark the PR as draft: [GitHub Docs](https://github.blog/2019-02-14-introducing-draft-pull-requests/)

This prevents merging of the PR before it's ready.

You should be able to mark it as draft retroactively in the sidebar on the right, if IIRC.

File: docs/style-properties.md
Comment (Weibye) [inline docs/style-properties.md:3]: This is a really good addition!

File: src/style.rs
Comment (Weibye) [inline src/style.rs:514]: We should try to get the flexbox gap implemented shortly after this or alongside of it

File: src/style.rs
Comment (Weibye) [inline src/style.rs:562]: Should this be split out into a separate type? 

Would it make sense to have a `FlexboxLayout` and a different `GridLayout`?

Comment (nicoburns) [reply src/style.rs:562]: If we really want to split it out, I imagine something like below would make sense. I worry that adding static typing constraints just makes thing unwieldy though. Especially as I imagine it will be rare for Taffy to be used directly by hand, and much more common for it to be translated from the surface syntax of another library/system. I wonder if it might not be better to be permission with what we accept, but print diagnostic information about ignored styles (or return it to the caller in a machine inspectable format).

```rust
enum InnerDisplay {
    Flex(FlexContainerStyle),
    Grid(GridContainerStyle),
    Leaf,
}

struct FlexContainerStyle {
    flex_direction: FlexDirection,
    flex_wrap: FlexWrap,

    align_content: AlignContent,
    align_items: AlignItems,
    justify_content: JustifyContent,
    justify_items: JustifyItems,
    gap: Size<Dimension>,
}


struct GridContainerStyle {
    grid_template_rows: GridTrackVec<GridTrackSizingFunction>,
    grid_template_columns: GridTrackVec<GridTrackSizingFunction>,
    grid_auto_rows: GridTrackVec<GridTrackSizingFunction>,
    grid_auto_columns: GridTrackVec<GridTrackSizingFunction>,
    grid_auto_flow: GridAutoFlow,

    align_content: AlignContent,
    align_items: AlignItems,
    justify_content: JustifyContent,
    justify_items: JustifyItems,
    gap: Size<Dimension>,
}

enum OuterDisplay {
    None,
    Root,
    FlexChild(FlexChildStyle),
    GridChild(GridChildStyle),
}

struct FlexChildStyle {
    flex_basis: Dimension,
    flex_grow: f32,
    flex_shrink: f32,
}

struct GridChildStyle {
    grid_row_start: GridLine,
    grid_row_end: GridLine,
    grid_column_start: GridLine,
    grid_column_end: GridLine,
}

struct CssStyle {
    /// What layout strategy should be used?
    outer_display: OuterDisplay,
    inner_display: InnerDisplay,

    // Position properties
    position_type: PositionType,
    position: Rect<Dimension>,

    // Size properies
    size: Size<Dimension>,
    min_size: Size<Dimension>,
    max_size: Size<Dimension>,
    aspect_ratio: Option<f32>,
    margin: Rect<Dimension>,
    padding: Rect<Dimension>,
    border: Rect<Dimension>,
}
```

File: src/style.rs
Comment (nicoburns) [reply src/style.rs:514]: Agreed. Flexbox implementation for `justify_items` and `justify_self` is also missing (see: https://github.com/DioxusLabs/taffy/issues/204#issuecomment-1193202234)

File: src/style.rs
Comment (alice-i-cecile) [reply src/style.rs:562]: This code base is hard enough to maintain as it is; I'm alright to increase the static typing to make it harder for us to have bugs.

Like you say, most end users will be using taffy in a wrapped form.

Comment (Weibye) [reply src/style.rs:562]: I think this design approach can make sense and in the direction I was picturing. 
It might be that we should wait until we have a 3rd layouting algorithm to make sure that what we build is general across the correct domain.

@alice-i-cecile any opinion?

Comment (alice-i-cecile) [reply src/style.rs:562]: This is the direction I was picturing too. I'd like to avoid adding `InnerDisplay` and `OuterDisplay` for this PR; it's a lot of complexity and can be nicely severed.

**Comment by alice-i-cecile** [comment]
As a heads up @nicoburns, I'd like to merge this relatively aggressively, then iterating on it in follow-up PRs before the next major version. Great work so far!

It's a much easier workflow for collaboration than trying to get it perfect in the very first PR.

**Comment by nicoburns** [comment]
@alice-i-cecile Good to know! I can break this up into smaller PRs if that would be helpful? And maybe add feature flags for flexbox and grid (flexbox defaulting to on, grid to off for now)?

However this PR builds on top of https://github.com/DioxusLabs/taffy/pull/202 (as it already much of the work of nicely factoring out the existing flexbox implementation into it's own module). I'm not sure what the status on that PR is, but currently this one would be blocked on that one being merged.

File: src/style.rs
Comment (nicoburns) [reply src/style.rs]: I now think we ought to do something similar to https://github.com/bevyengine/bevy/pull/5513, but I agree that it ought to be a separate PR (probably landing after this one, as it'll be easier to work out the correct splits for the styles once we have all of them).

**Comment by alice-i-cecile** [comment]
Awesome! I'm wildly impressed with your work here.

What type of review would be most useful to you?

File: src/axis.rs
Comment (alice-i-cecile) [inline src/axis.rs:1]: Should we move the flexbox axes (i.e. FlexDirection) into this file too?

File: docs/style-properties.md
Comment (alice-i-cecile) [inline docs/style-properties.md:1]: I love this file. Should we update what's implemented?

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
This needs release notes, ideally with a migration guide for the small changes needed to existing code.

File: src/style/dimension.rs
Comment (alice-i-cecile) [inline src/style/dimension.rs:78]: This doesn't seem distinct from `LengthPercentageAuto`? This name is way nicer though!

File: src/style/grid.rs
Comment (alice-i-cecile) [inline src/style/grid.rs:11]:
```suggestion
///
/// The "dense" packing algorithm attempts to fill in holes earlier in the grid, if smaller items come up later. This may cause items to appear out-of-order, when doing so would fill in holes left by larger items.
```

Line breaks will lead to nicer tooltips in IDEs.

File: src/style/grid.rs
Comment (alice-i-cecile) [inline src/style/grid.rs:55]:
```suggestion
/// A track placement specification. Used for grid-[row/column]-[start/end]. Named tracks are not implemented.
```

File: src/style/grid.rs
Comment (alice-i-cecile) [inline src/style/grid.rs:78]:
```suggestion
    /// Apply a mapping function if the [`GridPlacement`] is a `Track`. Otherwise return `self` unmodified.
```

File: src/style/grid.rs
Comment (alice-i-cecile) [inline src/style/grid.rs:150]: Does this docstring belong here? It needs more context if so.

File: src/style/grid.rs
Comment (alice-i-cecile) [inline src/style/grid.rs:157]: This comment both seems wrong and unhelpful.

File: src/style/grid.rs
Comment (alice-i-cecile) [inline src/style/grid.rs:248]: Ditto here on this description being unhelpful.

Also, line breaks :)

File: src/style/grid.rs
Comment (alice-i-cecile) [inline src/style/grid.rs:288]:
```suggestion
    /// will if the available_space is `Some`). Otherwise returns `None`.
```

File: src/style/grid.rs
Comment (alice-i-cecile) [inline src/style/grid.rs:356]:
```suggestion
/// Represents the type of automatic repetition to perform
```

Comment (alice-i-cecile) [inline src/style/grid.rs:356]: This could use a bit more explanation.

**Comment by nicoburns** [comment]
> What type of review would be most useful to you?

I'm not sure. I realise that any kind of review is going to be difficult due to the size of the PR. There are a few areas that I think it would definitely be useful to get feedback on:

- The refactor of the `style` module (and addition of grid styles) + additional helpers in the `style_helpers` module
- The `compute::common::alignment` module. This module contains alignment logic which is common between Flexbox and Grid. Do we think that makes sense? Or should I try to keep them separate?
- The usage of the [grid](https://github.com/becheran/grid) crate (it provides basically a 2D Vec on top of an actual Vec). We could implement something similar as an internal module, but I figured it might make more sense to use an external implementation given that's it's pretty generic functionality. This crate isn't particularly widely used (I couldn't find one that was that didn't bring in a whole bunch of linear algebra math we don't need), but it's small enough that we could easily take on the maintenance if we needed to.

> My instincts are to merge this pretty aggressively, to... reduce the tooling pain that comes with massive PRs.

Yeah, I agree. I think correctness/robustness for this code will likely be achieved with a comprehensive test suite (which we might need to build up over time) rather than code review. And once we have this one merged we can follow up with much smaller PRs.

> to let users test it ASAP 

On this note, I figure it might be a good idea to do alpha `0.3` releases where we make it clear that the Grid implementation isn't ready for production yet. I think it'll take some time to squash all the bugs, and it would be nice if people could try it before it's 100% ready. Perhaps they'll even send us bug reports.

File: src/style/dimension.rs
Comment (nicoburns) [reply src/style/dimension.rs:78]: It's not currently, but properties using `Dimension` support additional variants like `MinContent`, `MaxContent`, and `FitContent` that we don't support yet (but will probably want to at some point). As of the next release I wouldn't expect people to be using these types directly in any case. There was discussion of this on https://github.com/DioxusLabs/taffy/pull/269 where these types were introduced.

File: src/axis.rs
Comment (nicoburns) [reply src/axis.rs:1]: We could do. I've been trying to keep the `flex` and `grid` stuff separate in case we want to split it into separate crates and/or feature flag flexbox as well as grid. Currently everything in `axis` is CSS-specific, but not specific to any particular layout algorithm (CSS Grid doesn't have any non-standard axes).

**Comment by alice-i-cecile** [comment]
> The refactor of the style module (and addition of grid styles) + additional helpers in the style_helpers module

I like this!

> The compute::common::alignment module. This module contains alignment logic which is common between Flexbox and Grid. Do we think that makes sense? Or should I try to keep them separate?

This seems reasonable, but the exact details didn't feel quite right? I'd love if we could lump it in with other common functionality somewhere.

> The usage of the [grid](https://github.com/becheran/grid) crate (it provides basically a 2D Vec on top of an actual Vec). We could implement something similar as an internal module, but I figured it might make more sense to use an external implementation given that's it's pretty generic functionality. This crate isn't particularly widely used (I couldn't find one that was that didn't bring in a whole bunch of linear algebra math we don't need), but it's small enough that we could easily take on the maintenance if we needed to.

I'm kinda neutral on this. I'm fine with it for now.

> I figure it might be a good idea to do alpha 0.3 releases where we make it clear that the Grid implementation isn't ready for production yet

Yep I'd be happy with that!

**Comment by nicoburns** [comment]
@alice-i-cecile

I believe I have addressed all of your concerns/requests (in particular I've filled in the release notes). I have also:

- Implemented absolute and hidden positioning
- Fixed support for children with margins (including absolutely positioned children).
- Added a whole bunch of tests and fixes. We're up to 100 gentests for grid now (compared to 277 for Flexbox). Still a lot more testing to be done, but I think this is a good start.
- Updated CI to run tests both with and without the grid feature flag. Unfortunately the only way I could get this to work was remove it from the "taffy" section in Cargo.toml, which means that my editor doesn't seem to pick up grid related files by default, `cargo test` doesn't include grid tests by default, and clippy doesn't lint grid files by default (you have to specify `--features experimental_grid`). Not sure if there is anything we can do about this. For now I've been adding it back it temporarily but not committing it. Perhaps we should make it a default features? Not sure.

As always, progress is being tracked over at https://github.com/DioxusLabs/taffy/issues/204. My suggestion is that we try to get this PR merged ASAP, but that we keep the issue open and continue to use it to track overall progress on Grid. I imagine the release notes might need some work, but I would like to suggest we start a new release issue/PR and don't block this PR on that.

**Review by alice-i-cecile — APPROVED** [review]
Awesome write-up! That was my last blocking request: I think that merging and iterating is the way to go now.

**Pull request merged**
